From bf808730aad08b1052db9d2b91fdf79713886db4 Mon Sep 17 00:00:00 2001
From: John Ralls <jralls@ceridwen.us>
Date: Sun, 26 Oct 2014 15:39:21 -0700
Subject: [PATCH] Fix various issues when compiling natively on MinGW.

* On Windows, utf8.h and UTF8.h are the same, so #include <utf8.h> gets
UTF8.h. Rename the latter to WTF-UTF8.h
* Like Solaris, Windows needs to include either sys/time.h or time.h.
* The ICU I18N library is called libicuin instead of libicu18n.
* DumpRenderTree refused to link until the LDADD and LDFLAGS parameters
were removed from the makefile. Don't understand why, it just happened
to work when I was trying to get an error message.
---
 Source/JavaScriptCore/API/JSClassRef.cpp           |  2 +-
 Source/JavaScriptCore/API/JSStringRef.cpp          |  2 +-
 .../runtime/JSGlobalObjectFunctions.cpp            |  2 +-
 Source/JavaScriptCore/runtime/UString.cpp          |  2 +-
 Source/JavaScriptCore/tools/CodeProfiling.cpp      |  2 +-
 Source/JavaScriptCore/wtf/text/AtomicString.cpp    |  2 +-
 Source/JavaScriptCore/wtf/text/WTFString.cpp       |  2 +-
 Source/JavaScriptCore/wtf/unicode/UTF8.cpp         |  2 +-
 Source/JavaScriptCore/wtf/unicode/UTF8.h           | 84 ----------------------
 Source/JavaScriptCore/wtf/unicode/WTF-UTF8.h       | 84 ++++++++++++++++++++++
 .../wtf/unicode/glib/UnicodeGLib.cpp               |  2 +-
 Source/WebCore/loader/icon/IconRecord.h            |  2 +-
 Source/WebCore/page/Page.h                         |  2 +-
 .../platform/network/ResourceResponseBase.h        |  2 +-
 Source/WebCore/platform/text/String.cpp            |  2 +-
 Source/WebCore/xml/XSLTProcessorLibxslt.cpp        |  2 +-
 Source/WebKit2/Shared/WebString.h                  |  2 +-
 Tools/GNUmakefile.am                               | 34 ++++-----
 configure.ac                                       |  4 +-
 19 files changed, 118 insertions(+), 118 deletions(-)
 delete mode 100644 Source/JavaScriptCore/wtf/unicode/UTF8.h
 create mode 100644 Source/JavaScriptCore/wtf/unicode/WTF-UTF8.h

diff --git a/Source/JavaScriptCore/API/JSClassRef.cpp b/Source/JavaScriptCore/API/JSClassRef.cpp
index 298c734..d375a9d 100644
--- a/Source/JavaScriptCore/API/JSClassRef.cpp
+++ b/Source/JavaScriptCore/API/JSClassRef.cpp
@@ -34,7 +34,7 @@
 #include <runtime/ObjectPrototype.h>
 #include <runtime/Identifier.h>
 #include <wtf/text/StringHash.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 using namespace std;
 using namespace JSC;
diff --git a/Source/JavaScriptCore/API/JSStringRef.cpp b/Source/JavaScriptCore/API/JSStringRef.cpp
index ea31da6..1acb197 100644
--- a/Source/JavaScriptCore/API/JSStringRef.cpp
+++ b/Source/JavaScriptCore/API/JSStringRef.cpp
@@ -28,7 +28,7 @@
 
 #include "InitializeThreading.h"
 #include "OpaqueJSString.h"
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 using namespace JSC;
 using namespace WTF::Unicode;
diff --git a/Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp b/Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp
index 204144c..ade2997 100644
--- a/Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp
+++ b/Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp
@@ -42,7 +42,7 @@
 #include <wtf/Assertions.h>
 #include <wtf/MathExtras.h>
 #include <wtf/StringExtras.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 using namespace WTF;
 using namespace Unicode;
diff --git a/Source/JavaScriptCore/runtime/UString.cpp b/Source/JavaScriptCore/runtime/UString.cpp
index 5b1e9a0..52e697e 100644
--- a/Source/JavaScriptCore/runtime/UString.cpp
+++ b/Source/JavaScriptCore/runtime/UString.cpp
@@ -39,7 +39,7 @@
 #include <wtf/StringExtras.h>
 #include <wtf/Vector.h>
 #include <wtf/dtoa.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 #if HAVE(STRINGS_H)
 #include <strings.h>
diff --git a/Source/JavaScriptCore/tools/CodeProfiling.cpp b/Source/JavaScriptCore/tools/CodeProfiling.cpp
index f116038..1017caf 100644
--- a/Source/JavaScriptCore/tools/CodeProfiling.cpp
+++ b/Source/JavaScriptCore/tools/CodeProfiling.cpp
@@ -33,7 +33,7 @@
 #include <signal.h>
 #endif
 
-#if OS(LINUX)
+#if OS(LINUX) || defined(__MINGW32__)
 #include <sys/time.h>
 #endif
 
diff --git a/Source/JavaScriptCore/wtf/text/AtomicString.cpp b/Source/JavaScriptCore/wtf/text/AtomicString.cpp
index d775e7b..f2ef1ab 100644
--- a/Source/JavaScriptCore/wtf/text/AtomicString.cpp
+++ b/Source/JavaScriptCore/wtf/text/AtomicString.cpp
@@ -28,7 +28,7 @@
 #include <wtf/HashSet.h>
 #include <wtf/Threading.h>
 #include <wtf/WTFThreadData.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 namespace WTF {
 
diff --git a/Source/JavaScriptCore/wtf/text/WTFString.cpp b/Source/JavaScriptCore/wtf/text/WTFString.cpp
index 04c970a..2c1eba5 100644
--- a/Source/JavaScriptCore/wtf/text/WTFString.cpp
+++ b/Source/JavaScriptCore/wtf/text/WTFString.cpp
@@ -30,7 +30,7 @@
 #include <wtf/StringExtras.h>
 #include <wtf/Vector.h>
 #include <wtf/dtoa.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 #include <wtf/unicode/Unicode.h>
 
 using namespace std;
diff --git a/Source/JavaScriptCore/wtf/unicode/UTF8.cpp b/Source/JavaScriptCore/wtf/unicode/UTF8.cpp
index 8ea5c69..e0366da 100644
--- a/Source/JavaScriptCore/wtf/unicode/UTF8.cpp
+++ b/Source/JavaScriptCore/wtf/unicode/UTF8.cpp
@@ -25,7 +25,7 @@
  */
 
 #include "config.h"
-#include "UTF8.h"
+#include "WTF-UTF8.h"
 
 #include "ASCIICType.h"
 #include <wtf/StringHasher.h>
diff --git a/Source/JavaScriptCore/wtf/unicode/UTF8.h b/Source/JavaScriptCore/wtf/unicode/UTF8.h
deleted file mode 100644
index bc2497b..0000000
--- a/Source/JavaScriptCore/wtf/unicode/UTF8.h
+++ /dev/null
@@ -1,84 +0,0 @@
-/*
- * Copyright (C) 2007 Apple Inc.  All rights reserved.
- *
- * Redistribution and use in source and binary forms, with or without
- * modification, are permitted provided that the following conditions
- * are met:
- * 1. Redistributions of source code must retain the above copyright
- *    notice, this list of conditions and the following disclaimer.
- * 2. Redistributions in binary form must reproduce the above copyright
- *    notice, this list of conditions and the following disclaimer in the
- *    documentation and/or other materials provided with the distribution.
- *
- * THIS SOFTWARE IS PROVIDED BY APPLE COMPUTER, INC. ``AS IS'' AND ANY
- * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
- * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
- * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE COMPUTER, INC. OR
- * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
- * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
- * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
- * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
- * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
- * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
- */
-
-#ifndef WTF_UTF8_h
-#define WTF_UTF8_h
-
-#include "Unicode.h"
-
-namespace WTF {
-namespace Unicode {
-
-    // Given a first byte, gives the length of the UTF-8 sequence it begins.
-    // Returns 0 for bytes that are not legal starts of UTF-8 sequences.
-    // Only allows sequences of up to 4 bytes, since that works for all Unicode characters (U-00000000 to U-0010FFFF).
-    int UTF8SequenceLength(char);
-
-    // Takes a null-terminated C-style string with a UTF-8 sequence in it and converts it to a character.
-    // Only allows Unicode characters (U-00000000 to U-0010FFFF).
-    // Returns -1 if the sequence is not valid (including presence of extra bytes).
-    int decodeUTF8Sequence(const char*);
-
-    typedef enum {
-            conversionOK,       // conversion successful
-            sourceExhausted,    // partial character in source, but hit end
-            targetExhausted,    // insuff. room in target for conversion
-            sourceIllegal       // source sequence is illegal/malformed
-    } ConversionResult;
-
-    // These conversion functions take a "strict" argument. When this
-    // flag is set to strict, both irregular sequences and isolated surrogates
-    // will cause an error.  When the flag is set to lenient, both irregular
-    // sequences and isolated surrogates are converted.
-    // 
-    // Whether the flag is strict or lenient, all illegal sequences will cause
-    // an error return. This includes sequences such as: <F4 90 80 80>, <C0 80>,
-    // or <A0> in UTF-8, and values above 0x10FFFF in UTF-32. Conformant code
-    // must check for illegal sequences.
-    // 
-    // When the flag is set to lenient, characters over 0x10FFFF are converted
-    // to the replacement character; otherwise (when the flag is set to strict)
-    // they constitute an error.
-
-    WTF_EXPORT_PRIVATE ConversionResult convertUTF8ToUTF16(
-                    const char** sourceStart, const char* sourceEnd, 
-                    UChar** targetStart, UChar* targetEnd, bool strict = true);
-
-    ConversionResult convertLatin1ToUTF8(
-                    const LChar** sourceStart, const LChar* sourceEnd, 
-                    char** targetStart, char* targetEnd);
-
-    WTF_EXPORT_PRIVATE ConversionResult convertUTF16ToUTF8(
-                    const UChar** sourceStart, const UChar* sourceEnd, 
-                    char** targetStart, char* targetEnd, bool strict = true);
-
-    unsigned calculateStringHashAndLengthFromUTF8(const char* data, const char* dataEnd, unsigned& dataLength, unsigned& utf16Length);
-
-    bool equalUTF16WithUTF8(const UChar* a, const UChar* aEnd, const char* b, const char* bEnd);
-
-} // namespace Unicode
-} // namespace WTF
-
-#endif // WTF_UTF8_h
diff --git a/Source/JavaScriptCore/wtf/unicode/WTF-UTF8.h b/Source/JavaScriptCore/wtf/unicode/WTF-UTF8.h
new file mode 100644
index 0000000..bc2497b
--- /dev/null
+++ b/Source/JavaScriptCore/wtf/unicode/WTF-UTF8.h
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2007 Apple Inc.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE COMPUTER, INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE COMPUTER, INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
+ */
+
+#ifndef WTF_UTF8_h
+#define WTF_UTF8_h
+
+#include "Unicode.h"
+
+namespace WTF {
+namespace Unicode {
+
+    // Given a first byte, gives the length of the UTF-8 sequence it begins.
+    // Returns 0 for bytes that are not legal starts of UTF-8 sequences.
+    // Only allows sequences of up to 4 bytes, since that works for all Unicode characters (U-00000000 to U-0010FFFF).
+    int UTF8SequenceLength(char);
+
+    // Takes a null-terminated C-style string with a UTF-8 sequence in it and converts it to a character.
+    // Only allows Unicode characters (U-00000000 to U-0010FFFF).
+    // Returns -1 if the sequence is not valid (including presence of extra bytes).
+    int decodeUTF8Sequence(const char*);
+
+    typedef enum {
+            conversionOK,       // conversion successful
+            sourceExhausted,    // partial character in source, but hit end
+            targetExhausted,    // insuff. room in target for conversion
+            sourceIllegal       // source sequence is illegal/malformed
+    } ConversionResult;
+
+    // These conversion functions take a "strict" argument. When this
+    // flag is set to strict, both irregular sequences and isolated surrogates
+    // will cause an error.  When the flag is set to lenient, both irregular
+    // sequences and isolated surrogates are converted.
+    // 
+    // Whether the flag is strict or lenient, all illegal sequences will cause
+    // an error return. This includes sequences such as: <F4 90 80 80>, <C0 80>,
+    // or <A0> in UTF-8, and values above 0x10FFFF in UTF-32. Conformant code
+    // must check for illegal sequences.
+    // 
+    // When the flag is set to lenient, characters over 0x10FFFF are converted
+    // to the replacement character; otherwise (when the flag is set to strict)
+    // they constitute an error.
+
+    WTF_EXPORT_PRIVATE ConversionResult convertUTF8ToUTF16(
+                    const char** sourceStart, const char* sourceEnd, 
+                    UChar** targetStart, UChar* targetEnd, bool strict = true);
+
+    ConversionResult convertLatin1ToUTF8(
+                    const LChar** sourceStart, const LChar* sourceEnd, 
+                    char** targetStart, char* targetEnd);
+
+    WTF_EXPORT_PRIVATE ConversionResult convertUTF16ToUTF8(
+                    const UChar** sourceStart, const UChar* sourceEnd, 
+                    char** targetStart, char* targetEnd, bool strict = true);
+
+    unsigned calculateStringHashAndLengthFromUTF8(const char* data, const char* dataEnd, unsigned& dataLength, unsigned& utf16Length);
+
+    bool equalUTF16WithUTF8(const UChar* a, const UChar* aEnd, const char* b, const char* bEnd);
+
+} // namespace Unicode
+} // namespace WTF
+
+#endif // WTF_UTF8_h
diff --git a/Source/JavaScriptCore/wtf/unicode/glib/UnicodeGLib.cpp b/Source/JavaScriptCore/wtf/unicode/glib/UnicodeGLib.cpp
index a01c3ee..a99cefb 100644
--- a/Source/JavaScriptCore/wtf/unicode/glib/UnicodeGLib.cpp
+++ b/Source/JavaScriptCore/wtf/unicode/glib/UnicodeGLib.cpp
@@ -24,7 +24,7 @@
 #include "UnicodeGLib.h"
 
 #include <wtf/Vector.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 #define UTF8_IS_SURROGATE(character) (character >= 0x10000 && character <= 0x10FFFF)
 
diff --git a/Source/WebCore/loader/icon/IconRecord.h b/Source/WebCore/loader/icon/IconRecord.h
index 50ef7f7..a4d7c20 100644
--- a/Source/WebCore/loader/icon/IconRecord.h
+++ b/Source/WebCore/loader/icon/IconRecord.h
@@ -38,7 +38,7 @@
 #include <wtf/OwnPtr.h>
 #include <wtf/text/StringHash.h>
 
-#if OS(SOLARIS)
+#if OS(SOLARIS) || defined (__MINGW32__)
 #include <sys/types.h> // For time_t structure.
 #endif
 
diff --git a/Source/WebCore/page/Page.h b/Source/WebCore/page/Page.h
index 090b42c..eba7ca1 100644
--- a/Source/WebCore/page/Page.h
+++ b/Source/WebCore/page/Page.h
@@ -34,7 +34,7 @@
 #include <wtf/HashSet.h>
 #include <wtf/Noncopyable.h>
 
-#if OS(SOLARIS)
+#if OS(SOLARIS) || defined (__MINGW32__)
 #include <sys/time.h> // For time_t structure.
 #endif
 
diff --git a/Source/WebCore/platform/network/ResourceResponseBase.h b/Source/WebCore/platform/network/ResourceResponseBase.h
index 3c61666..5e76d17 100644
--- a/Source/WebCore/platform/network/ResourceResponseBase.h
+++ b/Source/WebCore/platform/network/ResourceResponseBase.h
@@ -35,7 +35,7 @@
 #include <wtf/PassOwnPtr.h>
 #include <wtf/RefPtr.h>
 
-#if OS(SOLARIS)
+#if OS(SOLARIS) || defined (__MINGW32__)
 #include <sys/time.h> // For time_t structure.
 #endif
 
diff --git a/Source/WebCore/platform/text/String.cpp b/Source/WebCore/platform/text/String.cpp
index f2f8d2e..78905ec 100644
--- a/Source/WebCore/platform/text/String.cpp
+++ b/Source/WebCore/platform/text/String.cpp
@@ -24,7 +24,7 @@
 
 #include "SharedBuffer.h"
 #include "TextBreakIterator.h"
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 #include <wtf/unicode/Unicode.h>
 
 using namespace WTF;
diff --git a/Source/WebCore/xml/XSLTProcessorLibxslt.cpp b/Source/WebCore/xml/XSLTProcessorLibxslt.cpp
index 076a852..fe5787a 100644
--- a/Source/WebCore/xml/XSLTProcessorLibxslt.cpp
+++ b/Source/WebCore/xml/XSLTProcessorLibxslt.cpp
@@ -50,7 +50,7 @@
 #include <wtf/Vector.h>
 #include <wtf/text/CString.h>
 #include <wtf/text/StringBuffer.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 #if PLATFORM(MAC)
 #include "SoftLinking.h"
diff --git a/Source/WebKit2/Shared/WebString.h b/Source/WebKit2/Shared/WebString.h
index 18b3fc8..0f04d70 100644
--- a/Source/WebKit2/Shared/WebString.h
+++ b/Source/WebKit2/Shared/WebString.h
@@ -30,7 +30,7 @@
 #include <JavaScriptCore/JSStringRef.h>
 #include <wtf/PassRefPtr.h>
 #include <wtf/text/WTFString.h>
-#include <wtf/unicode/UTF8.h>
+#include <wtf/unicode/WTF-UTF8.h>
 
 namespace WebKit {
 
diff --git a/Tools/GNUmakefile.am b/Tools/GNUmakefile.am
index fe7a1ce..8b56ee0 100644
--- a/Tools/GNUmakefile.am
+++ b/Tools/GNUmakefile.am
@@ -156,23 +156,23 @@ Programs_DumpRenderTree_CFLAGS = \
 	$(GTK_CFLAGS) \
 	$(LIBSOUP_CFLAGS)
 
-Programs_DumpRenderTree_LDADD = \
-	libjavascriptcoregtk-@WEBKITGTK_API_MAJOR_VERSION@.@WEBKITGTK_API_MINOR_VERSION@.la \
-	libwebkitgtk-@WEBKITGTK_API_MAJOR_VERSION@.@WEBKITGTK_API_MINOR_VERSION@.la \
-	libWebCoreInternals.la \
-	$(GLOBALDEPS_LIBS) \
-	$(CAIRO_LIBS) \
-	$(GTK_LIBS) \
-	$(GLIB_LIBS) \
-	$(LIBSOUP_LIBS) \
-	$(FREETYPE_LIBS) \
-	$(WINMM_LIBS) \
-	$(XRENDER_LIBS) \
-	$(XT_LIBS)
-
-Programs_DumpRenderTree_LDFLAGS = \
-	-no-fast-install \
-	-no-install
+#Programs_DumpRenderTree_LDADD = \
+#	libjavascriptcoregtk-@WEBKITGTK_API_MAJOR_VERSION@.@WEBKITGTK_API_MINOR_VERSION@.la \
+#	libwebkitgtk-@WEBKITGTK_API_MAJOR_VERSION@.@WEBKITGTK_API_MINOR_VERSION@.la \
+#	libWebCoreInternals.la \
+#	$(GLOBALDEPS_LIBS) \
+#	$(CAIRO_LIBS) \
+#	$(GTK_LIBS) \
+#	$(GLIB_LIBS) \
+#	$(LIBSOUP_LIBS) \
+#	$(FREETYPE_LIBS) \
+#	$(WINMM_LIBS) \
+#	$(XRENDER_LIBS) \
+#	$(XT_LIBS)
+
+#Programs_DumpRenderTree_LDFLAGS = \
+#	-no-fast-install \
+#	-no-install
 
 # clean target
 CLEANFILES += \
diff --git a/configure.ac b/configure.ac
index 126ca85..99beb38 100644
--- a/configure.ac
+++ b/configure.ac
@@ -419,8 +419,8 @@ if test "$with_unicode_backend" = "icu"; then
             UNICODE_LIBS="-licucore"
             ;;
         *-*-mingw*)
-            UNICODE_CFLAGS=""
-            UNICODE_LIBS="-licui18n -licuuc"
+            UNICODE_CFLAGS="-I/c/gcdev/webkit/include"
+            UNICODE_LIBS="-licuin -licuuc"
             ;;
         *)
             AC_PATH_PROG(icu_config, icu-config, no)
-- 
1.9.4.msysgit.0

